components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java

```diff
@@ -203,15 +227,144 @@ Get ancestor organization ids (including itself) of a given organization. The li
+                // Since Branding is inherited from the parent org, removing the parent org displayName.
+                removeOrgDisplayNameFromBrandingPreference(resolvedBrandingPreference);
+            }
+            return resolvedBrandingPreference;
+        }
+
+        // No cache found. Start with current organization application branding.
+        Optional<BrandingPreference> brandingPreference =
+                getBrandingPreference(APPLICATION_TYPE, appId, locale, currentTenantDomain);
+        if (brandingPreference.isPresent()) {
+            return brandingPreference.get();
+        }
+
+        // No application branding found. Check current organization branding.
+        brandingPreference = getBrandingPreference(ORGANIZATION_TYPE, currentTenantDomain, locale,
+                currentTenantDomain);
+        if (brandingPreference.isPresent()) {
+            addAppBrandingToCache(appId, currentTenantDomain, null, currentTenantDomain, ORGANIZATION_TYPE);
+            return brandingPreference.get();
+        }
+
+        /*
+            It is not possible to resolve application branding further if the organization ID is null or
+            if the current tenant domain is super tenant since it is the root organization.
+         */
+        if (orgId == null || MultitenantConstants.SUPER_TENANT_DOMAIN_NAME.equals(currentTenantDomain)) {
+            // No branding found. Adding the same tenant domain to cache to avoid the resolving in the next run.
+            addAppBrandingToCache(appId, currentTenantDomain, appId, currentTenantDomain, APPLICATION_TYPE);
+            throw handleClientException(ERROR_CODE_BRANDING_PREFERENCE_NOT_EXISTS,
+                    appId, APPLICATION_TYPE, getTenantDomain());
+        }
+
+        try {
+            // Get the details of the parent organization and resolve the branding preferences.
+            Organization organization = organizationManager.getOrganization(orgId, false, false);
+            String parentOrgId = organization.getParent().getId();
+            // This is a root tenant.
+            if (StringUtils.isBlank(parentOrgId)) {
+                return getPreference(ORGANIZATION_TYPE, currentTenantDomain, locale,
+                        currentTenantDomain);
+            }
+
+            OrgApplicationManager orgApplicationManager =
+                    BrandingResolverComponentDataHolder.getInstance().getOrgApplicationManager();
+            String parentAppId = orgApplicationManager.getParentAppId(appId, orgId);
+            String parentTenantDomain = organizationManager.resolveTenantDomain(parentOrgId);
+            int parentDepthInHierarchy = organizationManager.getOrganizationDepthInHierarchy(parentOrgId);
+
+            // Get the minimum hierarchy depth that needs to be reached to resolve branding preference
+            int minHierarchyDepth = Utils.getSubOrgStartLevel() - 1;
+
+            while (parentDepthInHierarchy >= minHierarchyDepth) {
```

Comment:
Refactor according to https://github.com/wso2/product-is/issues/16846

Reply:
Addressed with https://github.com/wso2-extensions/identity-branding-preference-management/pull/40/commits/05fa6da03cb482ecbd514d7030744011bad7a98e